hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/TestReplicationAnnotation.java

```diff
@@ -0,0 +1,139 @@
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Set;
+
+/**
+ * Tests on {@link org.apache.hadoop.hdds.scm.metadata.Replicate}.
+ */
+public class TestReplicationAnnotation {
+  private SCMHAInvocationHandler scmhaInvocationHandler;
+  private ContainerStateManagerV2 mockCSM;
+
+  @Before
+  public void setup() {
+    SCMHAManager mock = MockSCMHAManager.getInstance(true);
+    mockCSM = new ContainerStateManagerV2() {
+      @Override
+      public boolean contains(HddsProtos.ContainerID containerID) {
+        return false;
+      }
+
+      @Override
+      public Set<ContainerID> getContainerIDs() {
+        return null;
+      }
+
+      @Override
+      public Set<ContainerID> getContainerIDs(
+          HddsProtos.LifeCycleState state) {
+        return null;
+      }
+
+      @Override
+      public ContainerInfo getContainer(HddsProtos.ContainerID id) {
+        return null;
+      }
+
+      @Override
+      public Set<ContainerReplica> getContainerReplicas(
+          HddsProtos.ContainerID id) {
+        return null;
+      }
+
+      @Override
+      public void updateContainerReplica(
+          HddsProtos.ContainerID id, ContainerReplica replica) {
+
+      }
+
+      @Override
+      public void removeContainerReplica(
+          HddsProtos.ContainerID id, ContainerReplica replica) {
+
+      }
+
+      @Override
+      public void addContainer(HddsProtos.ContainerInfoProto containerInfo)
+          throws IOException {
+
+      }
+
+      @Override
+      public void updateContainerState(
+          HddsProtos.ContainerID id, HddsProtos.LifeCycleEvent event)
+          throws IOException, InvalidStateTransitionException {
+
+      }
+
+      @Override
+      public void updateDeleteTransactionId(
+          Map<ContainerID, Long> deleteTransactionMap) throws IOException {
+
+      }
+
+      @Override
+      public ContainerInfo getMatchingContainer(
+          long size, String owner, PipelineID pipelineID,
+          NavigableSet<ContainerID> containerIDs) {
+        return null;
+      }
+
+      @Override
+      public void removeContainer(HddsProtos.ContainerID containerInfo)
+          throws IOException {
+      }
+
+      @Override
+      public void close() throws IOException {
+
+      }
+    };
+
+    scmhaInvocationHandler = new SCMHAInvocationHandler(
+        RequestType.CONTAINER, mockCSM, mock.getRatisServer());
+  }
+
+  @Test
+  public void testReplicateAnnotationBasic() throws Throwable {
+    // test whether this call will hit the Ratis based replication
+    // code path in SCMHAInvocationHandler. The invoke() can return means the
+    // request is handled properly thus response is successful. Expected
+    // result is null because the function returns nothing.
+    Object[] arguments = {HddsProtos.ContainerInfoProto.getDefaultInstance()};
+    Assert.assertEquals(null, scmhaInvocationHandler.invoke(new Object(),
```

Comment:
Hey Rui, I consider you have to use my anonymous version of SCMRatisServer, 
because of  
```

      @Override
      public SCMRatisResponse submitRequest(SCMRatisRequest request) throws IOException, ExecutionException, InterruptedException {
        throw new IOException("submitRequest is called.");
      }
```
, 
from the IOException
```
    ContainerStateManagerV2 proxy = (ContainerStateManagerV2) Proxy.newProxyInstance(
        SCMHAInvocationHandler.class.getClassLoader(),
        new Class<?>[]{ContainerStateManagerV2.class}, invocationHandler);

    try {
      proxy.addContainer(HddsProtos.ContainerInfoProto.getDefaultInstance());
      Assert.fail();
    } catch (IOException ignore) {

    }
```
we can know that `invokeRatis()` is called, due to the annotation.